### PR TITLE
Ensure protokube can connect to kube-apiserver before starting the sync loop

### PR DESCRIPTION
**Pull request by olemarkus**
This commit will have protokube eagerly retry to connect to kube-apiserver instead of the 1 minute sync loop.
Another benefit is that we know that tasks are run in order and that labels are always applied before channels etc

With very few samples, it looks like a cluster gets the DNS entry about 1-2 minutes earlier, which I think makes sense.

Ref #11050

### CODE REVIEW

File: protokube/pkg/protokube/kube_boot.go [hunk 1]
```diff
@@ -101,8 +102,23 @@ func (k *KubeBoot) Init(volumesProvider Volumes) {
 
 // RunSyncLoop is responsible for provision the cluster
 func (k *KubeBoot) RunSyncLoop() {
+	ctx := context.Background()
+
+	for {
+		client, err := k.Kubernetes.KubernetesClient()
+		if err != nil {
+			klog.Warningf("could not create kubernetes client: %v", err)
```
Comment (justinsb) [inline protokube/pkg/protokube/kube_boot.go:110]: This one probably needs a sleep and continue?  Or klog.Fatalf?  I think we'll panic on the `client.CoreV1()` call anyway..

File: protokube/pkg/protokube/kube_boot.go [hunk 2]
```diff
@@ -101,8 +102,23 @@ func (k *KubeBoot) Init(volumesProvider Volumes) {
...
+		}
+
+		_, err = client.CoreV1().Namespaces().Get(ctx, "kube-system", metav1.GetOptions{})
+		if err == nil {
+			klog.Info("connected to the apiserver")
+			break
+		}
+		klog.Infof("Could not connect to apiserver: %v", err)
```
Comment (justinsb) [inline protokube/pkg/protokube/kube_boot.go:118]: Nit: something to indicate that this is normal(ish) might be helpful here.  e.g. "polling for apiserver readiness; not yet ready: %v" or something like that?

**Comment by justinsb** [comment]
This is a nice win.  LGTM; one error handling comment and one nit.

/approve
/lgtm

/hold

Feel free to remove the hold if you don't think the nits are worth it :-)

File: protokube/pkg/protokube/kube_boot.go (hunk 1 above)
Comment (olemarkus) [reply protokube/pkg/protokube/kube_boot.go:110]: We may as well panic here. There isn't much we can do if this one fails.

**Comment by rifelpet** [comment]
FWIW, Looking at the protokube logs from the first commit's [kubernetes-aws](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kops/11093/pull-kops-e2e-kubernetes-aws/1373576840320389120/artifacts/ip-172-20-60-121.ap-northeast-2.compute.internal/journal.log) and [k8s-containerd](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kops/11093/pull-kops-e2e-k8s-containerd/1373576840349749248/artifacts/ip-172-20-34-189.ap-northeast-2.compute.internal/journal.log) jobs it seems we retried 19 and 22 times respectively. We'll see how the most recent jobs perform as well.

**Review by rifelpet — APPROVED** [review]
most recent jobs both retried 15 times (75 seconds)

**Comment by justinsb** [comment]
Looks good, going to remove the hold (I had meant it to mean "feel free to remove hold when you want it to merge")

/hold cancel

**Pull request merged**